### PR TITLE
SI-8219 Avoid unpositioned ambiguous implicit error

### PR DESCRIPTION
**Pull request by retronym**
And characterization of the curiously case of overloaded ==

Review by @lrytz

### CODE REVIEW

File: src/compiler/scala/tools/nsc/typechecker/Typers.scala [hunk 1]
```diff
@@ -191,7 +191,7 @@ trait Typers extends Adaptations with Tags with TypersTracking with PatternTyper
          !from.isError
       && !to.isError
       && context.implicitsEnabled
-      && (inferView(EmptyTree, from, to, reportAmbiguous = false) != EmptyTree)
+      && silent(_.inferView(EmptyTree, from, to, reportAmbiguous = false)).fold(false)(_ != EmptyTree) // SI-8219 silent mode avoids leaking ambiguous errors
```
Comment (lrytz) [inline src/compiler/scala/tools/nsc/typechecker/Typers.scala]: Isn't it a bug in `inferImplicit`, in that it doesn't respect `reportAmbiguous = false`?

Comment (retronym) [inline src/compiler/scala/tools/nsc/typechecker/Typers.scala]: It doesn't report them immediately.

```
% scalac-hash v2.11.0-M7 -Ylog:typer -Ydebug test/files/pos/t8219.scala 2>&1 | grep update
[log typer] update buffer: Vector([Type error at:NoPosition] type mismatch;
```

``` scala
    if (result.isFailure && saveAmbiguousDivergent && implicitSearchContext.hasErrors) {
      context.updateBuffer(implicitSearchContext.reportBuffer.errors.collect {
        case dte: DivergentImplicitTypeError => dte
        case ate: AmbiguousImplicitTypeError => ate
      })
      debuglog("update buffer: " + implicitSearchContext.reportBuffer.errors)
    }
```

I'm going to try a slightly different approach with `saveErrors = false`.

Comment (retronym) [inline src/compiler/scala/tools/nsc/typechecker/Typers.scala]: pushed

**Comment by lrytz** [comment]
LGTM!

**Comment by retronym** [comment]
Hmm, looks like I'm walking a fine line here:

```
% cat test/files/pos/t8219b.scala
trait Arr[T]
object Arr {
  def apply[T](xs: T*)(implicit evidence$2: scala.reflect.ClassTag[T]): Arr[T] = ???
  def apply(x: Boolean,xs: Boolean*): Arr[Boolean]= ???
  def apply(x: Byte,xs: Byte*): Arr[Byte]= ???
  def apply(x: Short,xs: Short*): Arr[Short]= ???
  def apply(x: Char,xs: Char*): Arr[Char]= ???
  def apply(x: Int,xs: Int*): Arr[Int]= ???
  def apply(x: Long,xs: Long*): Arr[Long]= ???
  def apply(x: Float,xs: Float*): Arr[Float]= ???
  def apply(x: Double,xs: Double*): Arr[Double] = ???
  def apply(x: Unit,xs: Unit*): Arr[Unit] = ???
}

object Test1 {
  implicit def arrToTrav[T](a: Arr[T]): Traversable[T] = ???

  def foo(t: Traversable[Any]) {}

  Arr("")        // okay
  foo(Arr(""))   // fails 2.10.3, ticket/8129


  Array("")      // okay
  foo(Array("")) // okay 2.10.3, fails ticket/8129
}

% qbin/scalac test/files/pos/t8219b.scala
test/files/pos/t8219b.scala:21: error: overloaded method value apply with alternatives:
  (x: Unit,xs: Unit*)Arr[Unit] <and>
  (x: Double,xs: Double*)Arr[Double] <and>
  (x: Float,xs: Float*)Arr[Float] <and>
  (x: Long,xs: Long*)Arr[Long] <and>
  (x: Int,xs: Int*)Arr[Int] <and>
  (x: Char,xs: Char*)Arr[Char] <and>
  (x: Short,xs: Short*)Arr[Short] <and>
  (x: Byte,xs: Byte*)Arr[Byte] <and>
  (x: Boolean,xs: Boolean*)Arr[Boolean]
 cannot be applied to (String)
  foo(Arr(""))   // fails 2.10.3, ticket/8129
      ^
test/files/pos/t8219b.scala:25: error: overloaded method value apply with alternatives:
  (x: Unit,xs: Unit*)Array[Unit] <and>
  (x: Double,xs: Double*)Array[Double] <and>
  (x: Float,xs: Float*)Array[Float] <and>
  (x: Long,xs: Long*)Array[Long] <and>
  (x: Int,xs: Int*)Array[Int] <and>
  (x: Char,xs: Char*)Array[Char] <and>
  (x: Short,xs: Short*)Array[Short] <and>
  (x: Byte,xs: Byte*)Array[Byte] <and>
  (x: Boolean,xs: Boolean*)Array[Boolean]
 cannot be applied to (String)
  foo(Array("")) // okay 2.10.3, fails ticket/8129
      ^
two errors found

% scalac-hash v2.10.3 test/files/pos/t8219b.scala
test/files/pos/t8219b.scala:21: error: overloaded method value apply with alternatives:
  (x: Unit,xs: Unit*)Arr[Unit] <and>
  (x: Double,xs: Double*)Arr[Double] <and>
  (x: Float,xs: Float*)Arr[Float] <and>
  (x: Long,xs: Long*)Arr[Long] <and>
  (x: Int,xs: Int*)Arr[Int] <and>
  (x: Char,xs: Char*)Arr[Char] <and>
  (x: Short,xs: Short*)Arr[Short] <and>
  (x: Byte,xs: Byte*)Arr[Byte] <and>
  (x: Boolean,xs: Boolean*)Arr[Boolean]
 cannot be applied to (String)
  foo(Arr(""))   // fails 2.10.3, ticket/8129
      ^
one error found
```

**Comment by lrytz** [comment]
the `2.10.3` version looks dubious.

**Comment by retronym** [comment]
During `isApplicable` for the generic apply method, we hit an ambiguous error when determining of the result type, `Array[?T]` is coercible to `Traversable[Any]`. The difference between `Arr` and `Arr` are the implicits in scope `Predef.{wrapLongArray, .. }`. More than one of these are deemed potentially applicable. 

Anyway, we rely on that ambiguity bubbling up into the outer context used in `Infer#tryTwice`, which resolves things on the second pass with by disabling views. Or maybe its the caller to `tryTwice` which attempts to recover from failure with wildcard expected type.

**Comment by retronym** [comment]
Here's a slightly smaller minimization. 

``` scala
trait Arr[T]
object Arr {
  def apply[T](xs: T*)        : Arr[T]    = ???
  def apply(x: Long,xs: Long*): Arr[Long] = ???
}

object I {
  implicit def arrToTrav[T] (a: Arr[T])   : Traversable[T]    = ???  
  implicit def longArrToTrav(a: Arr[Long]): Traversable[Long] = ???
}

object Test {
  def foo(t: Traversable[Any]) {}

  object Okay {
    Arr("")

    import I.{ arrToTrav, longArrToTrav }
    foo(Arr(""))
  }

  object Fail {
    import I.arrToTrav
    foo(Arr("")) // 2.10.3: found String, expected Long
  }
}

```

**Comment by retronym** [comment]
Lodged as [SI-8230](https://issues.scala-lang.org/browse/SI-8230).

I need to think a bit more about what to do for this bug. It's tempting to peel off the last commit, but I don't think its safe until we get to the bottom of SI-8230. Closing for now.

**Pull request closed without merging**